KAFKA-14499: [6/N] Add MemberId and MemberEpoch to OffsetFetchRequest

dajac opened the pull request:
This patch adds the MemberId and the MemberEpoch fields to the OffsetFetchRequest. Those fields will be populated when the new consumer group protocol is used to ensure that the member fetching the offset has the correct member id and epoch. If it does not, UNKNOWN_MEMBER_ID or STALE_MEMBER_EPOCH are returned to the client.

Our initial idea was to implement the same for the old protocol. The field is called GenerationIdOrMemberEpoch in KIP-848 to materialize this. As a second though, I think that we should only do it for the new protocol. The effort to implement it in the old protocol is not worth it in my opinion.

This patch adds unit test to test the new code. I will add integration tests in a separate PR.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dajac committed:
KAFKA-14499: [6/N] Add MemberId and MemberEpoch to OffsetFetchRequest

rreddy-22 commented on clients/src/main/resources/common/message/OffsetFetchRequest.json:
```diff
@@ -47,8 +54,12 @@
+      { "name": "GroupId", "type": "string", "versions": "8+", "entityType": "groupId",
         "about": "The group ID."},
+      { "name": "MemberId", "type": "string", "versions": "9+", "ignorable": true,
+        "about": "The member ID assigned by the group coordinator if using the consumer protocol." },
```
should we specify new consumer group protocol?

CalvinLiu7947 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -538,17 +537,46 @@ public void validateOffsetCommit(
+        long lastCommittedOffset
+    ) throws UnknownMemberIdException, StaleMemberEpochException {
+        // When the member epoch is -1, the request comes from the admin client. In this case,
+        // the request can commit offsets if the group is empty.
```
Should we check the group is empty here? The code seems misaligned with the comment.

jolshan replied:
I think the "when the member epoch is -1" part applies here. Is the empty group part somewhere else?

jolshan replied:
I made a comment on something similar in another PR.  Might be a slightly different case, but similar wording about committing offsets is allowed and it being a bit unclear if fetching them is or isn't

https://github.com/apache/kafka/pull/14120/files/7e5f1f64d173d3e3b68fd6310fb76d6137daed2c#r1303310001

dajac replied:
The comment is wrong. Let me fix it.

CalvinLiu7947 commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/OffsetMetadataManagerTest.java:
```diff
@@ -1409,6 +1444,82 @@ public void testFetchAllOffsetsAtDifferentCommittedOffset() {
+    }
+
+    @Test
+    public void testConsumerGroupOffsetFetchWithStaleMemberEpoch() {
```
Maybe add the case for admin client?

jolshan replied:
This is the member epoch -1 right?

jolshan replied:
And to be clear -- this is the group has -1 and we are asking for 10.

the case Calvin is asking about seems to be passing -1 in.

dajac replied:
Sure.

CalvinLiu7947 reviewed (commented):
Thanks for the PR!

jolshan commented on clients/src/main/resources/common/message/OffsetFetchResponse.json:
```diff
@@ -32,8 +32,21 @@
+  // - COORDINATOR_NOT_AVAILABLE (version 0+)
+  // - COORDINATOR_LOAD_IN_PROGRESS (version 0+)
+  // - GROUP_ID_NOT_FOUND (version 0+)
+  // - UNSTABLE_OFFSET_COMMIT (version 7+)
```
thanks for adding this block -- it's good to have such a mapping on the json for responses.

jolshan commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java:
```diff
@@ -225,17 +225,21 @@ private Group validateOffsetCommit(
 
     /**
-     * Validates an OffsetCommit request.
+     * Validates an OffsetFetch request.
```
whoops -- was this a typo before 😅

jolshan commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroup.java:
```diff
@@ -823,9 +823,17 @@ public void validateOffsetCommit(
     @Override
-    public void validateOffsetFetch() throws GroupIdNotFoundException {
+    public void validateOffsetFetch(
+        String memberId,
```
For my understanding, we pass these to be consistent with the interface, but we don't need it?

dajac replied:
Correct. I mentioned this in the javadoc of the method.

jolshan commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/OffsetMetadataManagerTest.java:
```diff
@@ -181,10 +181,28 @@ public List<OffsetFetchResponseData.OffsetFetchResponseTopics> fetchOffsets(
             OffsetFetchResponseData.OffsetFetchResponseGroup response = offsetMetadataManager.fetchOffsets(
                 new OffsetFetchRequestData.OffsetFetchRequestGroup()
                     .setGroupId(groupId)
+                    .setMemberId(memberId)
```
Is this needed now because of validation (and wasn't needed before)

dajac replied:
Correct.

jolshan commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/OffsetMetadataManagerTest.java:
```diff
@@ -181,10 +181,28 @@ public List<OffsetFetchResponseData.OffsetFetchResponseTopics> fetchOffsets(
             String groupId,
             List<OffsetFetchRequestData.OffsetFetchRequestTopics> topics,
             long committedOffset
+        ) {
```
is this for generic groups only now?

dajac replied:
This or to simulate the admin client case.

jolshan replied:
It could be useful to add a small comment in case folks wanted to build upon these tests. But up to you.

jolshan commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroupTest.java:
```diff
@@ -631,4 +631,33 @@ public void testValidateOffsetCommit() {
+        ConsumerGroup group = new ConsumerGroup(snapshotRegistry, "group-foo");
+
+        // Simulate a call from the admin client without member id and member epoch.
+        // This should pass only if the group is empty.
```
Do we have an example where it fails because the group is not empty?

dajac replied:
That part of the comment was wrong. I just removed it.

jolshan commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroupTest.java:
```diff
@@ -631,4 +631,33 @@ public void testValidateOffsetCommit() {
+        snapshotRegistry.getOrCreateSnapshot(0);
+        group.getOrMaybeCreateMember("member-id", true);
+
+        // The member does not exist at epoch 0.
```
Should this be offset 0? Epoch 0 seems to be correct in the last line of the test

dajac replied:
This comment is confusing. The epoch was referring to the epoch of the snapshot here. I rephrased it to make it clearer.

dajac committed:
fix test

dajac committed:
fix tests

dajac committed:
address comments

dajac commented:
@rreddy-22 @CalvinConfluent @jolshan Thanks for your comments. I have addressed all of them. I have also changed a bit the schema and fixed a few other tests.

CalvinLiu7947 reviewed (approved):
LGTM

jolshan commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/OffsetMetadataManagerTest.java:
```diff
@@ -1491,10 +1526,15 @@ public void testConsumerGroupOffsetFetchWithUnknownMemberId() {
 
+        // Fetch offsets cases.
+        assertThrows(UnknownMemberIdException.class,
+            () -> context.fetchOffsets("group", "", 0, topics, Long.MAX_VALUE));
```
Is this testing that the empty member string will fail (ie if no member, it must be null?) I'm trying to understand how this test is different than the one below.

I also noticed that the test changed the default values from empty string to null. Was that an error before?

jolshan replied:
Or is this part of the latest commit that made null the no member ID indicator?

dajac replied:
Yeah, I changed the member id to be nullable and null by default. I think that it makes more sense. So now, an empty string raises an unknown member id error.

jolshan reviewed (approved):
Thanks David :)
